### DISCUSSION TITLE
Laggy stream with WebRTC and adaptive streaming enabled

### DISCUSSION THREAD

**Discussion by radosdaniel** [ant-media/Ant-Media-Server#5120] [Q&A] [answered]
Hey,

When I turn on adaptive streaming, and stream with WebRTC (using the test WebRTCApp) the stream gets very laggy compared to streaming with adaptive streaming off.
We tested it on a brand new instance on AWS, all default configurations.

On an other server we tried several configurations, but nothing helped.
Streaming with RTMP works fine, but WebRTC is very important for our product.

Thanks for your help!

**Comment by tremzo** [comment]
Are you just testing with one live stream and what server are you using? 

Does it look like system resources are getting really high with adaptive bitrates enabled? Typically, as soon as you start using ABR, the system resources do increase and an increase in resources.

**Reply by radosdaniel** [reply]
Yes, we are testing with one live stream only. It's a c5.xlarge instance (which is the default option)
The CPU load is between 15-25% which is, so I don't think that's a problem.

**Reply by tremzo** [reply]
What Ant Media version are you using? 2.5.3?

**Reply by radosdaniel** [reply]
Yes, It's Enterprise Edition 2.5.3 20230107_0851

**Reply by tremzo** [reply]
@radosdaniel, can you provide any demo page so we can see the lagging you are experiencing?   Then we will be able to troubleshoot the issue from there.

**Reply by radosdaniel** [reply]
You can start a stream here: https://media2.fitters.live/WebRTCAppEE/ (it accepts undefined streams for now).
You can check the stream here with the correct stream id: https://media2.fitters.live/WebRTCAppEE/play.html?name=STREAM_ID

Just to let you know, we are going to shut down this instance on Monday, because we created this one for this testing only.

Thanks!

**Reply by tremzo** [reply]
I set the max bitrate on the publishing page to 3000 and forced 1080p on the playback and did not experience any lagging or buffering. Although I did when I left all settings on the default. Did you experience something different?

**Reply by pestig** [reply]
How can I force 1080p on the player site ( https://media2.fitters.live/WebRTCAppEE/play.html?id=...), I can't find option for that.
@timantmedia

**Reply by tremzo** [reply]
If you have adaptive bitrate enabled, then click the options button and a new button/dropdown will appear with "Force Stream Resolution"

<img width="767" alt="Screenshot 2023-05-22 at 09 54 16" src="https://github.com/ant-media/Ant-Media-Server/assets/121940452/1cf09e0e-ee37-46c9-99f7-6ead2c382a70">

**Reply by mekya** [reply]
Hi @radosdaniel,

Maybe the problem is about the encoding parameters of the RTMP stream. Because RTMP encoders may add delay to compress better. 

Please make sure that there is no B-frames and encoder is tuned for zerolatency(which is a parameter in OBS - x264)

If you still have this problem, let us know. We can schedule a meeting to understand your problem and do our best to resolve. 

Regards
Oguz

**Reply by radosdaniel** [reply]
Hi,

Thanks for your help, we tried streaming with a max of 3000 bitrate and forced the resolution as you advised, but sadly still not looking any better.
What I discovered is that if I select 720p and 480p for the adaptive bitrates, it looks smooth. Once I set it back to 720p and 1080p the stream is choppy again.
I also tried to change the instance to c5.2xlarge, but it won't help.

RTMP streaming seems good, we only have this issue once streaming with WebRTC.

I would like to schedule a meeting to this Friday if it is possible.

Thanks!

**Reply by mekya** [reply]
I see. If you use c5.xlarge first, it may not be powerful enough to transcode 1080p streams. Check the CPU usage. 

Or you may be pointing out [this issue](https://github.com/ant-media/Ant-Media-Server/issues/4471) that occurs in 1080p bitrates, or there may be something different in your case. Anyway let's figure it out. 

Friday at 5pm or 6pm GMT+3 works for you?

**Reply by radosdaniel** [reply]
We tried it on a c5.2xlarge instance as well with the same result (CPU usage on c5.xlarge is maxed around 25%).
We will try out the new version, anyway I would like to schedule a meeting at Friday 5pm!

Thank you!

**Reply by tremzo** [reply]
Which version are you using? Are you using the latest Ant Media Server version?

**Reply by radosdaniel** [reply]
Now we are using Enterprise Edition 2.5.3 20230107_0851.

**Reply by mekya** [reply]
Here is the meeting at Friday 5pm!
Check this out 

https://calendar.google.com/calendar/event?action=TEMPLATE&tmeid=MTZkaWdzYTJsYXVnamozaDlyMXM3cXBrc3YgbWVreWFAYW50bWVkaWEuaW8&tmsrc=mekya%40antmedia.io

**Reply by radosdaniel** [reply]
Thank you!

Unfortunately I can't accept the invitation, I think there's something wrong with the link.
Can you send me a new invitation?

Thanks!

**Reply by mekya** [reply]
My pleasure. I hope the below button helps

<a target="_blank" href="https://calendar.google.com/calendar/event?action=TEMPLATE&amp;tmeid=MTZkaWdzYTJsYXVnamozaDlyMXM3cXBrc3YgbWVreWFAYW50bWVkaWEuaW8&amp;tmsrc=mekya%40antmedia.io"><img border="0" src="https://www.google.com/calendar/images/ext/gc_button1_en.gif"></a>


If not, here is the info

Meeting with Radosdaniel and Ant Media
Friday, May 26 · 5:00 – 5:45pm
Time zone: Europe/Istanbul
Google Meet joining info
Video call link: https://meet.google.com/imc-jpnh-smm
Or dial: ‪(GB) +44 20 3956 0905‬ PIN: ‪939 657 731‬#
More phone numbers: https://tel.meet/imc-jpnh-smm?pin=9138541230410

**Reply by mekya** [reply]
Hi @radosdaniel,
Are you joining the meeting? 

Regards
Oguz

**Reply by mekya** [reply]
Hi @radosdaniel,
Thank you for joining the meeting. Let me give the meeting minutes.

Version 2.6.0
- Publishing with WebRTC on sample page
- With single bitrate, when the bitrate is the default value for 1080p and 720p on c5.xlarge, it plays smooths both in 1080p and 720p
- With two bitrates(1080p - 3Mbps, 720p- 2Mbps), the playback is not smooth and we monitor that it's not smooth on webrtc-internals 

Diagnosis:
c5.xlarge is not fast enough to encode the two bitrates in real-time on the other hand CPU load is about %30-40. There should be some thread parameters for openh264 to increase threading. By default, thread number is 2x CPU cores so it's 8 for your case.
Quick workaround solution is trying GPU instance or some other faster CPU instances. 
Permanent solution is I'm just going to take a look if there are some parameters that may help improve the smoothness.

Let me know when you've some results for the GPU instance.

Have a great weekend
Oguz

**Reply by radosdaniel** [reply]
Hey @mekya,

We made some testing after we talked, and in the end we decided to move forward with a GPU instance from the AWS marketplace.
Can you help us with transferring our current Ant Media license to this construction?

Thank you!

**Reply by yashtandon113** [reply]
Hi @radosdaniel 
If you would like to use marketplace image of Ant Media Server with GPU so you need to subscribe to the GPU image of AMS.
You can check [here](https://aws.amazon.com/marketplace/pp/prodview-mn2mmjdggepse?sr=0-4&ref_=beagle&applicationId=AWSMPContessa).

**Reply by mekya** [reply]
Hi guys,
Thank you for your answer @yashtandon113. I think @radosdaniel question is more likely about license transferring. Please correct me if I'm wrong @radosdaniel  

Just let me explain it further. 
- If you just use AMS GPU from AWS Marketplace, it has its own license no need to transfer any license. You can stop your previous instance or cancel your subscription to antmedia.io if you have a license from antmedia.io 

- If you have a license from antmedia.io and would like to use it on an AWS GPU instance, you need to install Ant Media Server to a GPU instance and use the license directly. No need for any transfer because your license is floating and can be used anywhere.
If you have difficulty with installation, let us know. We can help you with the installation

Please let us know if we can do anything to help and proceed.

Cheers

**Reply by tremzo** [reply]
@mekya we have a FD99739 with same issue from same customer with same issue on GPU instance. Did you manage to find recommended parametres?

**Reply by mekya** [reply]
Generally, there is no problem for GPU instance in this case. The problem is that CPU is not powerful to encoder in real-time and when we use GPU, it's working well. I think the problem is something else in that ticket.

**Reply by tremzo** [reply, marked as answer]
@mekya I found the customer application setting encoderPreset: slower. I think this is the issue. I changed this preset and tried various combinations on the OBS settings to achieve a better quality of playback. https://github.com/ant-media/Ant-Media-Server/issues/5464#issuecomment-1679461177
